7_python_flask_fastapi/lesson1/app_fastapi.py

```diff
@@ -0,0 +1,217 @@
+    sqlalchemy.Column("order_status", sqlalchemy.String(32)),
+)
+
+engine = sqlalchemy.create_engine(DATABASE_URL)
+metadata.create_all(engine)
+
+app = FastAPI()
+
+class UserIn(BaseModel):
+    first_name: str = Field(max_length=32)
+    last_name: str = Field(max_length=32)
+    email: str = Field(max_length=128)
+    password: str = Field(max_length=128)
+
+class User(BaseModel):
+    id: int
+    first_name: str = Field(max_length=32)
+    last_name: str = Field(max_length=32)
+    email: str = Field(max_length=128)
+
+class ProductIn(BaseModel):
+    name: str = Field(max_length=64)
+    description: str = Field(max_length=256)
+    price: float
+
+class Product(BaseModel):
+    id: int
+    name: str = Field(max_length=64)
+    description: str = Field(max_length=256)
+    price: float
+
+class orderIn(BaseModel):
+    user_id: int
+    product_id: int
+    order_date: Optional[datetime] = None
+    order_status: Optional[str] = None
+
+class order(BaseModel):
+    id: int
+    user_id: int
+    product_id: int
+    order_date: Optional[datetime] = None
+    order_status: Optional[str] = None
+
+@app.post("/users/", response_model=User)
+async def create_user(user: UserIn):
+    query = users.insert().values(
+        first_name=user.first_name,
+        last_name=user.last_name,
+        email=user.email,
+        password=user.password,
+    )
+    last_record_id = await database.execute(query)
+    return {**user.dict(), "id": last_record_id}
+
+@app.get("/users/", response_model=List[User])
+async def read_users():
+    query = users.select()
+    return await database.fetch_all(query)
+
+@app.get("/users/{user_id}", response_model=User)
+async def read_user(user_id: int):
+    query = users.select().where(users.c.id == user_id)
+    return await database.fetch_one(query)
+
+@app.put("/users/{user_id}", response_model=User)
+async def update_user(user_id: int, new_user: UserIn):
+    query = (
+        users.update()
+        .where(users.c.id == user_id)
+        .values(
+            first_name=new_user.first_name,
+            last_name=new_user.last_name,
+            email=new_user.email,
+            password=new_user.password,
+        )
+    )
+    await database.execute(query)
+    return {**new_user.dict(), "id": user_id}
+
+@app.delete("/users/{user_id}")
+async def delete_user(user_id: int):
+    query = users.delete().where(users.c.id == user_id)
+    await database.execute(query)
+    return {"message": "User deleted"}
+
+@app.post("/products/", response_model=Product)
+async def create_product(product: ProductIn):
+    query = products.insert().values(
+        name=product.name,
+        description=product.description,
+        price=product.price,
+    )
+    last_record_id = await database.execute(query)
+    return {**product.dict(), "id": last_record_id}
+
+@app.get("/products/", response_model=List[Product])
+async def read_products():
+    query = products.select()
+    return await database.fetch_all(query)
```

Comment:
Вот неплохой способ вывода табличной информации
```suggestion
@app.get("/products/", response_class=HTMLResponse)
async def read_products():
    query = products.select()
    data = await database.fetch_all(query)
    return HTMLResponse(pd.DataFrame(data).to_html())
```
для этого нужно добавить таки импорты
```python
import pandas as pd
from fastapi.responses import HTMLResponse
```
Команда для установки pandas: `pip install pandas`